### PR TITLE
Apply .NET8 `MethodInvoker`/`ConstructorInvoker` optimization

### PR DESCRIPTION
**Pull request by SergeiPavlov**
Described here:
https://github.com/dotnet/runtime/pull/88415
https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8#reflection

### CODE REVIEW

**Review by alex-kulakov — CHANGES_REQUESTED** [review]

File: Orm/Xtensive.Orm/IoC/ServiceContainer.cs [hunk 1]
```diff
@@ -106,10 +118,14 @@ protected virtual object CreateInstance(ServiceRegistration serviceInfo)
       finally {
         _ = creating.TryRemove(key, out _);
       }
+#if NET8_0_OR_GREATER
+      return cInfo.Invoke(new Span<object>(args));
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/IoC/ServiceContainer.cs]: done

File: Orm/Xtensive.Orm/Reflection/TypeHelper.cs [hunk 2]
```diff
@@ -53,7 +53,11 @@ public int GetHashCode((Type, Type[]) obj)
     private static readonly Type CompilerGeneratedAttributeType = typeof(CompilerGeneratedAttribute);
     private static readonly string TypeHelperNamespace = typeof(TypeHelper).Namespace;
 
+#if NET8_0_OR_GREATER
+    private static readonly ConcurrentDictionary<(Type, Type[]), ConstructorInvoker> ConstructorInfoByTypes =
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Reflection/TypeHelper.cs]: done

File: Orm/Xtensive.Orm/Reflection/TypeHelper.cs [hunk 3]
```diff
@@ -685,6 +714,7 @@ public static object Activate(this Type type, Type[] genericArguments, params ob
       return constructors.SingleOrDefault();
     };
 
+
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Reflection/TypeHelper.cs]: removed it

File: Orm/Xtensive.Orm/Reflection/TypeHelper.cs [hunk 4]
```diff
@@ -920,6 +950,20 @@ private static string InnerGetTypeName(this Type type, bool useShortForm)
...
+      GenericMethodInvokers1.GetOrAdd((genericDefinition, typeArgument), GenericMethodInvokerFactory1);
+
+    public static MethodInvoker CachedMakeGenericMethodInvoker(this MethodInfo genericDefinition, Type typeArgument1, Type typeArgument2) =>
+      GenericMethodInvokers2.GetOrAdd((genericDefinition, typeArgument1, typeArgument2), GenericMethodInvokerFactory2);
+#else
+    public static MethodInfo CachedMakeGenericMethodInvoker(this MethodInfo genericDefinition, Type typeArgument) =>
+      CachedMakeGenericMethod(genericDefinition, typeArgument);
+
+    public static MethodInfo CachedMakeGenericMethodInvoker(this MethodInfo genericDefinition, Type typeArgument1, Type typeArgument2) =>
+      CachedMakeGenericMethod(genericDefinition, typeArgument1, typeArgument2);
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Reflection/TypeHelper.cs]: done

**Review by alex-kulakov — APPROVED** [review]

**Comment by alex-kulakov** [comment]
@SergeiPavlov, I'm going to make new ```GetSingleConstructorInvokerOrDefault```/```GetSingleConstructorInvoker``` methods internal. Are you ok with that?

**Comment by SergeiPavlov** [comment]
I'm Ok, but what about `GetSingleConstructorOrDefault()` ? 
It is public but conditionally available only on <`.NET8`

**Comment by alex-kulakov** [comment]
Well, it is public and correct way of getting it out of public is to make it obsolete and then make it internal. It is a single-usage method. Probably, it will be deleted after NET6 and NET7 will be removed from project target frameworks. For now, all I can do is to make it obsolete to hint the future for possible users.

BTW, I'll make it exist in NET8 too to prevent possible compilation error on user side on framework upgrade, another reason is that ConstructorInvoker and ConstructorInfo are not interchangeable.

**Pull request merged**